src/platform_impl/linux/wayland/seat/dnd/handlers.rs

```diff
@@ -0,0 +1,136 @@
+                        WindowEvent::CursorMoved {
+                            device_id: crate::event::DeviceId(
+                                crate::platform_impl::DeviceId::Wayland(DeviceId),
+                            ),
+                            position: PhysicalPosition::new(x, y),
+                            modifiers: Default::default(),
+                        },
+                        window_id,
+                    );
+
+                    for path in paths {
+                        winit_state
+                            .event_sink
+                            .push_window_event(WindowEvent::HoveredFile(path), window_id);
+                    }
+                    inner.window_id = Some(window_id);
+                }
+            }
+        }
+        DndEvent::Drop { offer: Some(offer) } => {
+            if let Some(window_id) = inner.window_id {
+                inner.window_id = None;
+
+                if let Ok(paths) = parse_offer(&winit_state.display, offer) {
+                    for path in paths {
+                        winit_state
+                            .event_sink
+                            .push_window_event(WindowEvent::DroppedFile(path), window_id);
+                    }
+                }
+            }
+        }
+        DndEvent::Leave => {
+            if let Some(window_id) = inner.window_id {
+                inner.window_id = None;
+
+                winit_state
+                    .event_sink
+                    .push_window_event(WindowEvent::HoveredFileCancelled, window_id);
+                winit_state.event_sink.push_window_event(
+                    WindowEvent::CursorLeft {
+                        device_id: crate::event::DeviceId(crate::platform_impl::DeviceId::Wayland(
+                            DeviceId,
+                        )),
+                    },
+                    window_id,
+                );
+            }
+        }
+        DndEvent::Motion { x, y, .. } => {
+            if let Some(window_id) = inner.window_id {
+                winit_state.event_sink.push_window_event(
+                    WindowEvent::CursorMoved {
+                        device_id: crate::event::DeviceId(crate::platform_impl::DeviceId::Wayland(
+                            DeviceId,
+                        )),
+                        position: PhysicalPosition::new(x, y),
+                        modifiers: Default::default(),
+                    },
+                    window_id,
+                );
+            }
+        }
+        _ => {}
+    }
+}
+
+fn parse_offer(display: &Display, offer: &DataOffer) -> io::Result<Vec<PathBuf>> {
+    let can_accept = offer.with_mime_types(|types| types.iter().any(|s| s == MIME_TYPE));
+    if can_accept {
+        // Format: https://www.iana.org/assignments/media-types/text/uri-list
+        let mut paths = Vec::new();
+        let pipe = offer.receive(MIME_TYPE.into())?;
```

Comment:
The pipe must be inserted and handled via callop pipe reader and be non blocking when doing so, otherwise it can deadlock.